pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -323,13 +345,21 @@ func (ctrl *PersistentVolumeController) syncUnboundClaim(claim *v1.PersistentVol
 		} else /* pv != nil */ {
 			// Found a PV for this claim
 			// OBSERVATION: pvc is "Pending", pv is "Available"
-			klog.V(4).Infof("synchronizing unbound PersistentVolumeClaim[%s]: volume %q found: %s", claimToClaimKey(claim), volume.Name, getVolumeStatusForLogging(volume))
+			claimKey := claimToClaimKey(claim)
+			klog.V(4).Infof("synchronizing unbound PersistentVolumeClaim[%s]: volume %q found: %s", claimKey, volume.Name, getVolumeStatusForLogging(volume))
 			if err = ctrl.bind(volume, claim); err != nil {
 				// On any error saving the volume or the claim, subsequent
 				// syncClaim will finish the binding.
+				// record count error for provision if exists, do not clean up cache
+				metrics.RecordMetric(claimKey, &ctrl.operationTimestamps, err)
 				return err
 			}
 			// OBSERVATION: claim is "Bound", pv is "Bound"
+			// if exists a timestamp entry in cache, record end to end provision latency and clean up cache
+			// End of the provision + binding operation lifecycle
+			// [Unit test 12-1, 12-2, 12-4]
+			metrics.RecordMetric(claimKey, &ctrl.operationTimestamps, nil)
+			ctrl.operationTimestamps.Delete(claimKey)
```

Comment:
Do you think we could move `RecordMetric` as a function inside `operationTimestamps` - so as act of recording the metric also automatically deletes the key from cache? It will less error prone that way?

Reply:
agreed leaving the life-cycle management of a cache item to the caller, i.e., in this case pv_controller/pv_controller_base, may causes cache entry leaking if caller forgot to call Delete. I propose to manage the deletion within RecordMetric, however not letting that function inside OperationStartTimeCache but still under metrics.go instead. To me, the cache should not be dealing with cache entry lifecycle management. WDYT?

Reply:
updated "RecordMetric" as described above. when an operation has been successfully executed, i.e., err == nil, the corresponding cache'd entry will be removed within the function.